### ISSUE TITLE
Problems with decryption using AES_CBC_decrypt_buffer()

### ISSUE THREAD

**Issue by zenichev** [kokke/tiny-AES-c#134]
Hi there! I'm sorry  to disturb you for such a simple issue, but unfortunately I've come across one problem that prevents me decrypting messages in some sporadic way.

I have a pretty basic setup with: client - server topology.
Both sides have the same AES key and an initial vector for decryption/encryption functionality.

I'm working currently with 256 size of AES key, and CBC algorithm:
AES_CBC_decrypt_buffer();
AES_CBC_encrypt_buffer();

This works well for a while, and after some period of time it gets a sporadic issue with a failed decrypting on the client's side. And it can happen that after a failed decryption, next time it gets properly done one.

I'm aware that I have to use this: AES_ctx_set_iv() before each encrypting/decrypting, so to let an initial vector be the same for (encrypted row and for decryption of this), so I do.
I'm also aware that there is no padding for CBC and I have to use buffers multiple to 16, so I do.

What is much interesting, that a server in it's turn is able to decrypt the same row when client experiences an issue (I inserted a testing decryption on the server side, to see if it works at all), but client (having the same AES key and IV) is not able to do this (when experiencing the issue).

I noticed that, when this issue happens, an encrypted row is suspiciously short.
So when it works out well it has an encrypted row about 500-700 symbols.
And when this issues happens it has about 50-60 symbols only after encryption, so that's quite strange (and this 50-60 symbols row is able to be decrypted on the server's side, but how? ).

Thanks for your attention, and once again sorry for disturbing you!

**Comment by kokke** [comment]
Hi @zenichev and no worries about asking for help. I will close this issue, because it is more usage related than a bug-report, but do feel free to keep commenting (you should be able to even if the issue is closed).

To me it sounds like you are either padding incorrectly or trying to use ASCII as input - these are the two most common mistakes I see. Please confirm whether or not you are transmitting ASCII strings.

You say you are using buffers with sizes in multiples of 16 bytes - that's good. 
But unless you are always utilizing all bytes in the buffers, you will need a padding scheme. How do you handle that currently?

I will advise you to print out the key, IV, input, input length in bytes and the output on every encrypt/decrypt operation.
Then, if you can't decrypt on the client-side, but you can on the server-side, one of the parameters mentioned above should differ (key, iv etc.). Usually this is enough to spot where the error is.

Let me know if you make any progress :)

**Comment by kokke** [comment]
>'m also aware that there is no padding for CBC and I have to use buffers multiple to 16, so I do.

Please elaborate on the padding. What do you mean bu "no padding"? :)

**Comment by zenichev** [comment]
> To me it sounds like you are either padding incorrectly or trying to use ASCII as input - these are the two most common mistakes I see. Please confirm whether or not you are transmitting ASCII strings.

Should I fill the rest of an unused buffer by something, that will just fit a space up to the end of it?
That's my fault I think, I don't do any filling right now (of an unused buffer part).
I will try to show how currently my encryption process looks like, using this simplified example:

```
void someFunction(struct AES_ctx * ctx, uint8_t * IV, uint8_t * key) {

	uint8_t echoBuffer[1600]={0};										// declare a buffer muplitple to 16
	strcpy(echoBuffer, "someStringIwannaEncryptRightNowYadaYadaYada");	// copy some data into the buffer

	AES_ctx_set_iv(ctx, IV);											// re-initialize initial vector
	AES_CBC_encrypt_buffer(ctx, echoBuffer, sizeof(echoBuffer));		// encrypt formed string and write this in the same buffer

	// after that I can use this somehow, for e.g. sending to a remote side
}
```


> You say you are using buffers with sizes in multiples of 16 bytes - that's good.
> But unless you are always utilizing all bytes in the buffers, you will need a padding scheme. How do you handle that currently?

Yeah that's perhaps where my fault is, I don't do any padding as I understood, I thought the only thing I'm supposed to do - is to keep size of buffer multiple to 16.

> I will advise you to print out the key, IV, input, input length in bytes and the output on every encrypt/decrypt operation.

That's what I actually do when debugging (all that you mentioned), and the weird thing was, that key and IV were always staying the same on both sides, that's why I started to doubt about it.

**Comment by zenichev** [comment]
I came up with a pretty basic function for padding, that is:
```
// buffer - is uint8_t buffer with data
// bufferSize - is a full buffer size
// bufferLength - is a length of stored string
int fillBuffer(uint8_t * buffer, int bufferSize, int buferLength) {

	uint8_t * ptrSrc = buffer;
	uint8_t * paddingValue = "#";

	// how many array members to fill - padding
	// add "- 1" at the end not to override return carriage
	int amount = bufferSize - buferLength - 1;
	
	if ( amount < 0 ) return 0;	// quit if we are not able to calculate amount for padding

	while(*ptrSrc && buferLength-->0) ptrSrc++;		// move pointer up to the end of string

	for (int i=amount; i>0; i--) *ptrSrc++=*paddingValue;	//add padding to each null member

	return 0;
}
```
seems to be working, perhaps you have a better one? can you advise something? Thanks : )

**Comment by kokke** [comment]
You need to fill all the space in the buffer with something, even if you don't use it.
Otherwise you will risk including data from previous telegrams in new telegrams, if the new telegram is shorter than previously sent ones.

Example, first you send one telegram of length 16, consisting of 16 * 'x'. Buffer contains 16 * 'x'.
Then you send another telegram of length 10, consisting of 10 * 'y'. If you don't do something to clear the buffer before/after sending, or perform padding, the buffer will consist of 10 * 'y' and then 6 * 'x' because only the first 10 bytes have been overwritten.

Do you understand?

I don't have time to review your code now, but I will later.

**Comment by zenichev** [comment]
@kokke alright, I understood. I came up with a solution for padding of the buffer, and with a functionality that is able to clear out padding if needed (for decrypted rows).

Surely I clear buffers after each usage. Currently I'm working with threads, and each thread has its own buffer (that is cleared after processing).

Yes, I understood what you meant. Unfortunately even having padding filled for buffer to be encrypted, I came across this issue once again, I don't get what I'm doing wrong.

So my steps are:
1. create a uint8_t buffer (multiple to 16) and fill it with needed metadata.
2. add padding at the end of it (so to fill all buffer).
3. encrypt this buffer (with initialization of IV before encryption).
4. send to remote side.
5. remote side gets it and decrypts with a wrong data (sporadically, not always).

And again, when this issue occurs, encrypted data is suspiciously short, like 50-60 symbols (while buffer with padding has 1600 symbols of length before en encryption).

To clarify:
- I have the same key and IV, that I see when debugging values of them (on both sides).

**Comment by kokke** [comment]
> And again, when this issue occurs, encrypted data is suspiciously short, like 50-60 symbols (while buffer with padding has 1600 symbols of length before en encryption).

How do you determine the length when you receive something?
It seems from your code that you are always encrypting 1600 bytes and sending it.

**Comment by zenichev** [comment]
Well, in case with CBC, what I do is actually filling the rest of unused buffer with padding (currently I use a hash '#' symbol for doing this), and then perform an encryption, that's it.

I'm following the length of string using strlen() command in my prints, such as:
printf("DEBUG: something here yada yada: %d  \n", strlen(someCharArray));

So I monitor:
- a length of a real metadata before padding
- a length of completely filled buffer (before an encryption)
- a length of a buffer (after an encryption is done)

And for side that receives an encrypted row:
- a length of a gotten encrypted data (data is stored in uint8_t array)
- a length of decrypted data

Usually when my row is messed after decryption, it has part of the beginning that is properly decrypted and part that is just a series of non-sense symbols, like this:
"here_I_had_some_needed_inform&#IU#NKFEfmekfe1123e23r"

Currently I decided to switch to CTR, since there I don't have to deal with a multiple length of the buffer to 16, and, no need for padding. Currently this one is working fine.
I think that with case of CBC, I used a short buffer that wasn't capable to store all amount of an encrypted data, I had like 1600 length that was hardcoded. Perhaps it's reasonable to use dynamic length for buffers.
